File: sdks/ts/memory/src/ingest/image.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Image OCR extractor using PaddleOCR (primary) with Tesseract as
 * fallback. Both engines are invoked as subprocesses — no native
 * bindings required. Implements the Extractor interface from P1-5.
 */

import { mkdtemp, rm, writeFile } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import type { Readable } from 'node:stream'
import type {
  ExtractOptions,
  ExtractResult,
  Extractor,
  ExtractorCapability,
  MagicSignature,
} from './extractor.js'
import { DEFAULT_SUBPROCESS_TIMEOUT_MS, checkBinaryAvailable, runSubprocess } from './subprocess.js'

/** Configuration for the image OCR extractor. */
export type ImageExtractorConfig = {
  /** Path or name of the PaddleOCR binary. Default: "paddleocr". */
  readonly paddleOcrBinary?: string
  /** Path or name of the Tesseract binary. Default: "tesseract". */
  readonly tesseractBinary?: string
  /** Default OCR language hint. Default: "en". */
  readonly defaultLanguage?: string
  /** Subprocess timeout in milliseconds. Default: 60000. */
  readonly timeout?: number
}

/** Maps ISO 639-1 codes to PaddleOCR language parameters. */
const PADDLE_OCR_LANGUAGE_MAP: Readonly<Record<string, string>> = {
  en: 'en',
  zh: 'ch',
  ja: 'japan',
  ko: 'korean',
  fr: 'fr',
  de: 'german',
  es: 'es',
  pt: 'pt',
  it: 'it',
  ru: 'ru',
  ar: 'ar',
  hi: 'hi',
  ta: 'ta',
  te: 'te',
}

/** Maps ISO 639-1 codes to Tesseract 3-letter language codes. */
const TESSERACT_LANGUAGE_MAP: Readonly<Record<string, string>> = {
  en: 'eng',
  zh: 'chi_sim',
  ja: 'jpn',
  ko: 'kor',
  fr: 'fra',
  de: 'deu',
  es: 'spa',
  pt: 'por',
  it: 'ita',
  ru: 'rus',
  ar: 'ara',
  hi: 'hin',
  ta: 'tam',
  te: 'tel',
}

/** Maps MIME types to file extensions for temp files. */
const CONTENT_TYPE_EXTENSION_MAP: Readonly<Record<string, string>> = {
  'image/png': '.png',
  'image/jpeg': '.jpg',
  'image/tiff': '.tiff',
  'image/bmp': '.bmp',
  'image/webp': '.webp',
}

/** Image content types handled by this extractor. */
const IMAGE_CONTENT_TYPES: readonly string[] = [
  'image/png',
  'image/jpeg',
  'image/tiff',
  'image/bmp',
  'image/webp',
]

/** File extensions handled by this extractor. */
const IMAGE_EXTENSIONS: readonly string[] = [
  '.png',
  '.jpg',
  '.jpeg',
  '.tiff',
  '.tif',
  '.bmp',
  '.webp',
]

/** Magic byte signatures for supported image formats. */
const IMAGE_MAGIC_BYTES: readonly MagicSignature[] = [
  { offset: 0, bytes: new Uint8Array([0x89, 0x50, 0x4e, 0x47]) }, // PNG
  { offset: 0, bytes: new Uint8Array([0xff, 0xd8, 0xff]) }, // JPEG
  { offset: 0, bytes: new Uint8Array([0x49, 0x49, 0x2a, 0x00]) }, // TIFF LE
  { offset: 0, bytes: new Uint8Array([0x4d, 0x4d, 0x00, 0x2a]) }, // TIFF BE
  { offset: 0, bytes: new Uint8Array([0x42, 0x4d]) }, // BMP
  { offset: 0, bytes: new Uint8Array([0x52, 0x49, 0x46, 0x46]) }, // WebP (RIFF)
]

/**
 * Resolves a configuration value with env var override support.
 */
const envOrDefault = (envKey: string, fallback: string): string => {
  const value = process.env[envKey]
  return value !== undefined && value !== '' ? value : fallback
}

/**
 * Resolves the effective timeout from config and environment.
 */
const resolveTimeout = (configTimeout?: number): number => {
  if (configTimeout !== undefined && configTimeout > 0) return configTimeout
  const envMs = process.env.MEMORY_EXTRACTOR_TIMEOUT_MS
  if (envMs !== undefined && envMs !== '') {
    const parsed = Number.parseInt(envMs, 10)
    if (!Number.isNaN(parsed) && parsed > 0) return parsed
  }
  return DEFAULT_SUBPROCESS_TIMEOUT_MS
}

/**
 * Maps an ISO 639-1 language code to PaddleOCR's format.
 */
const mapLanguageToPaddleOCR = (lang: string): string => {
  const normalised = lang.trim().toLowerCase()
  return PADDLE_OCR_LANGUAGE_MAP[normalised] ?? normalised
}

/**
 * Maps an ISO 639-1 language code to Tesseract's 3-letter format.
 */
const mapLanguageToTesseract = (lang: string): string => {
  const normalised = lang.trim().toLowerCase()
  return TESSERACT_LANGUAGE_MAP[normalised] ?? 'eng'
}

/**
 * Determines the file extension from extract options.
 */
const extensionFromOpts = (opts: ExtractOptions): string => {
  const fromMime = CONTENT_TYPE_EXTENSION_MAP[opts.contentType]
  if (fromMime !== undefined) return fromMime

  if (opts.fileName !== undefined) {
    const dotIdx = opts.fileName.lastIndexOf('.')
    if (dotIdx >= 0) return opts.fileName.slice(dotIdx)
  }

  return '.png'
}

/**
 * Writes raw bytes to a secure temporary file and returns the path.
 * Creates a dedicated temp directory to prevent collisions.
 */
const writeTempFile = async (
  raw: Buffer,
  extension: string,
): Promise<{ path: string; dir: string }> => {
  const dir = await mkdtemp(join(tmpdir(), 'memory-ocr-'))
  const path = join(dir, `input${extension}`)
  await writeFile(path, raw)
  return { path, dir }
}

/**
 * Strips surrounding single or double quotes from a string.
 */
const stripQuotes = (s: string): string => {
  if (s.length < 2) return s
  const first = s[0]
  const last = s[s.length - 1]
  if ((first === "'" && last === "'") || (first === '"' && last === '"')) {
    return s.slice(1, -1)
  }
  return s
}

/** Parsed OCR result with text blocks and average confidence. */
type ParsedOCRResult = {
  readonly text: string
  readonly confidence: number
}

/**
 * Parses PaddleOCR output. Handles both JSON array format and
 * tuple-style line format.
 */
export const parsePaddleOCROutput = (output: string): ParsedOCRResult => {
  const lines = output.split('\n')
  const textBlocks: string[] = []
  let totalConfidence = 0
  let blockCount = 0

  for (const rawLine of lines) {
    const line = rawLine.trim()
    if (line === '') continue

    // Try JSON array format: [["text", confidence], ...]
    try {
      const entries = JSON.parse(line) as unknown
      if (Array.isArray(entries)) {
        for (const entry of entries) {
          if (Array.isArray(entry) && entry.length >= 2) {
            const text = entry[0]
            const conf = entry[1]
            if (typeof text === 'string' && typeof conf === 'number') {
              textBlocks.push(text)
              totalConfidence += conf
              blockCount++
            }
          }
        }
        continue
      }
    } catch {
      // Not JSON; try tuple format.
    }

    // Try tuple format: ('text', 0.95)
    if (line.startsWith('(') && line.endsWith(')')) {
      const inner = line.slice(1, -1)
      const lastComma = inner.lastIndexOf(',')
      if (lastComma > 0) {
        const textPart = stripQuotes(inner.slice(0, lastComma).trim())
        const confPart = inner.slice(lastComma + 1).trim()
        const conf = Number.parseFloat(confPart)
        if (!Number.isNaN(conf)) {
          textBlocks.push(textPart)
          totalConfidence += conf
          blockCount++
        }
      }
    }
  }

  return {
    text: textBlocks.join('\n'),
    confidence: blockCount > 0 ? totalConfidence / blockCount : 0,
  }
}

/**
 * Creates an ImageExtractor that extracts text from images using
 * PaddleOCR (primary) with Tesseract as fallback.
 */
export const createImageExtractor = (config?: ImageExtractorConfig): Extractor => {
  const paddleOcrBinary =
    config?.paddleOcrBinary ?? envOrDefault('MEMORY_PADDLEOCR_PATH', 'paddleocr')
  const tesseractBinary =
    config?.tesseractBinary ?? envOrDefault('MEMORY_TESSERACT_PATH', 'tesseract')
  const defaultLanguage = config?.defaultLanguage ?? 'en'
  const timeout = resolveTimeout(config?.timeout)

  const resolveLanguage = (opts: ExtractOptions): string => opts.encoding ?? defaultLanguage

  const extractWithPaddleOCR = async (
    raw: Buffer,
    language: string,
    opts: ExtractOptions,
    signal?: AbortSignal,
  ): Promise<ExtractResult> => {
    const ext = extensionFromOpts(opts)
    const { path: tmpPath, dir: tmpDir } = await writeTempFile(raw, ext)

    try {
      const paddleLang = mapLanguageToPaddleOCR(language)

      const result = await runSubprocess(
        paddleOcrBinary,
        ['--image_dir', tmpPath, '--use_angle_cls', 'true', '--lang', paddleLang, '--type', 'ocr'],
        undefined,
        { timeout, signal },
      )

      if (result.exitCode !== 0) {
        throw new Error(`paddleocr exited with code ${result.exitCode}: ${result.stderr}`)
      }

      const parsed = parsePaddleOCROutput(result.stdout.toString('utf-8'))

      return {
        text: parsed.text,
        contentType: opts.contentType,
        encoding: 'UTF-8',
        metadata: {
          ocr_engine: 'paddleocr',
          ocr_confidence: parsed.confidence.toFixed(4),
          ocr_language: paddleLang,
        },
        pages: 0,
        language,
        confidence: parsed.confidence,
        skipped: false,
      }
    } finally {
      await rm(tmpDir, { recursive: true, force: true }).catch(() => {})
    }
  }

  const extractWithTesseract = async (
    raw: Buffer,
    language: string,
    opts: ExtractOptions,
    signal?: AbortSignal,
  ): Promise<ExtractResult> => {
    const ext = extensionFromOpts(opts)
    const { path: tmpPath, dir: tmpDir } = await writeTempFile(raw, ext)

    try {
      const tesseractLang = mapLanguageToTesseract(language)

      const result = await runSubprocess(
        tesseractBinary,
        [tmpPath, 'stdout', '-l', tesseractLang, '--oem', '3', '--psm', '3'],
        undefined,
        { timeout, signal },
      )

      if (result.exitCode !== 0) {
        throw new Error(`tesseract exited with code ${result.exitCode}: ${result.stderr}`)
      }

      const text = result.stdout.toString('utf-8').trim()

      return {
        text,
        contentType: opts.contentType,
        encoding: 'UTF-8',
        metadata: {
          ocr_engine: 'tesseract',
          ocr_language: tesseractLang,
        },
        pages: 0,
        language,
        confidence: 0,
        skipped: false,
      }
    } finally {
      await rm(tmpDir, { recursive: true, force: true }).catch(() => {})
    }
  }

  const extractor: Extractor = {
    name: 'image-ocr',

    contentTypes: IMAGE_CONTENT_TYPES,

    async extract(raw: Buffer, opts: ExtractOptions, signal?: AbortSignal): Promise<ExtractResult> {
      if (raw.length === 0) {
        return {
          text: '',
          contentType: opts.contentType,
          encoding: '',
          metadata: { ocr_engine: 'none' },
          pages: 0,
          language: '',
          confidence: 0,
          skipped: true,
          reason: 'empty input',
        }
      }

      const language = resolveLanguage(opts)

      // Try PaddleOCR first.
      if (await checkBinaryAvailable(paddleOcrBinary)) {
        try {
          return await extractWithPaddleOCR(raw, language, opts, signal)
        } catch {
          // Fall through to Tesseract.
        }
      }

      // Fallback to Tesseract.
      if (await checkBinaryAvailable(tesseractBinary)) {
        try {
          return await extractWithTesseract(raw, language, opts, signal)
        } catch (err: unknown) {
          throw new Error(
            `ingest: image extraction failed with both OCR engines: ${err instanceof Error ? err.message : String(err)}`,
          )
        }
      }

      throw new Error(
        `ingest: no OCR engine available (tried ${paddleOcrBinary}, ${tesseractBinary})`,
      )
    },

    async extractStream(
      source: Readable,
      opts: ExtractOptions,
      signal?: AbortSignal,
    ): Promise<ExtractResult> {
      const chunks: Buffer[] = []
      let totalBytes = 0

      for await (const chunk of source) {
        const buf = Buffer.isBuffer(chunk)
          ? chunk
          : chunk instanceof Uint8Array
            ? Buffer.from(chunk)
            : Buffer.from(String(chunk))

        if (opts.maxBytes !== undefined && opts.maxBytes > 0) {
          const remaining = opts.maxBytes - totalBytes
          if (remaining <= 0) break
          chunks.push(buf.subarray(0, remaining))
          totalBytes += Math.min(buf.length, remaining)
          if (totalBytes >= opts.maxBytes) break
        } else {
          chunks.push(buf)
          totalBytes += buf.length
        }
      }

      return extractor.extract(Buffer.concat(chunks), opts, signal)
    },

    async available(): Promise<boolean> {
      const paddleAvailable = await checkBinaryAvailable(paddleOcrBinary)
      if (paddleAvailable) return true
      return checkBinaryAvailable(tesseractBinary)
    },

    capability(): ExtractorCapability {
      return {
        extensions: IMAGE_EXTENSIONS,
        mimeTypes: IMAGE_CONTENT_TYPES,
        magicBytes: IMAGE_MAGIC_BYTES,
        requiresBinary: true,
      }
    },
  }

  return extractor
}

File: go/ingest/image.go
// SPDX-License-Identifier: Apache-2.0

package ingest

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// ImageExtractorConfig configures the image OCR extractor.
type ImageExtractorConfig struct {
	// PaddleOCRBinary is the path or name of the PaddleOCR binary.
	// Default: "paddleocr". Override via MEMORY_PADDLEOCR_PATH env var.
	PaddleOCRBinary string

	// TesseractBinary is the path or name of the Tesseract binary.
	// Default: "tesseract". Override via MEMORY_TESSERACT_PATH env var.
	TesseractBinary string

	// DefaultLanguage is the OCR language hint. Default: "en".
	DefaultLanguage string

	// Timeout for each OCR subprocess invocation.
	// Default: 60 seconds. Override via MEMORY_EXTRACTOR_TIMEOUT_MS env var.
	Timeout time.Duration

	// Logger for diagnostic output. Uses slog.Default() when nil.
	Logger *slog.Logger
}

// applyDefaults fills zero-value fields with defaults and environment
// variable overrides.
func (c *ImageExtractorConfig) applyDefaults() {
	if c.PaddleOCRBinary == "" {
		c.PaddleOCRBinary = envOrDefault("MEMORY_PADDLEOCR_PATH", "paddleocr")
	}
	if c.TesseractBinary == "" {
		c.TesseractBinary = envOrDefault("MEMORY_TESSERACT_PATH", "tesseract")
	}
	if c.DefaultLanguage == "" {
		c.DefaultLanguage = "en"
	}
	if c.Timeout <= 0 {
		if ms := os.Getenv("MEMORY_EXTRACTOR_TIMEOUT_MS"); ms != "" {
			if v, err := strconv.ParseInt(ms, 10, 64); err == nil && v > 0 {
				c.Timeout = time.Duration(v) * time.Millisecond
			}
		}
		if c.Timeout <= 0 {
			c.Timeout = DefaultSubprocessTimeout
		}
	}
	if c.Logger == nil {
		c.Logger = slog.Default()
	}
}

// envOrDefault returns the value of the named environment variable, or
// the fallback when the variable is empty or unset.
func envOrDefault(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

// ImageExtractor extracts text from images using PaddleOCR (primary)
// with Tesseract as fallback. Both engines are invoked as subprocesses
// — no native bindings or cgo required.
type ImageExtractor struct {
	cfg ImageExtractorConfig
}

// Compile-time interface check.
var _ Extractor = (*ImageExtractor)(nil)

// NewImageExtractor creates an ImageExtractor with the given config.
// Zero-value fields are populated with sensible defaults.
func NewImageExtractor(cfg ImageExtractorConfig) *ImageExtractor {
	cfg.applyDefaults()
	return &ImageExtractor{cfg: cfg}
}

// imageContentTypes lists all MIME types handled by the image extractor.
var imageContentTypes = []string{
	"image/png",
	"image/jpeg",
	"image/tiff",
	"image/bmp",
	"image/webp",
}

// imageExtensions lists all file extensions handled by the image extractor.
var imageExtensions = []string{
	".png", ".jpg", ".jpeg", ".tiff", ".tif", ".bmp", ".webp",
}

// imageMagicBytes contains magic byte signatures for supported image
// formats.
var imageMagicBytes = []MagicSignature{
	{Offset: 0, Bytes: []byte{0x89, 0x50, 0x4E, 0x47}},             // PNG
	{Offset: 0, Bytes: []byte{0xFF, 0xD8, 0xFF}},                    // JPEG
	{Offset: 0, Bytes: []byte{0x49, 0x49, 0x2A, 0x00}},              // TIFF (little-endian)
	{Offset: 0, Bytes: []byte{0x4D, 0x4D, 0x00, 0x2A}},              // TIFF (big-endian)
	{Offset: 0, Bytes: []byte{0x42, 0x4D}},                          // BMP
	{Offset: 0, Bytes: []byte{0x52, 0x49, 0x46, 0x46}},              // WebP (RIFF header)
}

// Name returns the extractor identifier.
func (e *ImageExtractor) Name() string {
	return "image-ocr"
}

// ContentTypes returns the MIME types this extractor handles.
func (e *ImageExtractor) ContentTypes() []string {
	return imageContentTypes
}

// Capability returns the capability descriptor for image extraction.
func (e *ImageExtractor) Capability() ExtractorCapability {
	return ExtractorCapability{
		Extensions:     imageExtensions,
		MIMETypes:      imageContentTypes,
		MagicBytes:     imageMagicBytes,
		RequiresBinary: true,
	}
}

// Available reports whether at least one OCR engine (PaddleOCR or
// Tesseract) is installed on the system. Returns false gracefully
// when neither is available.
func (e *ImageExtractor) Available(ctx context.Context) (bool, error) {
	paddleAvailable := CheckBinaryAvailable(ctx, e.cfg.PaddleOCRBinary)
	if paddleAvailable {
		return true, nil
	}

	tesseractAvailable := CheckBinaryAvailable(ctx, e.cfg.TesseractBinary)
	return tesseractAvailable, nil
}

// Extract performs OCR on the raw image bytes. PaddleOCR is attempted
// first; if unavailable or failing, Tesseract is used as fallback.
// Returns a descriptive error when both engines are unavailable.
func (e *ImageExtractor) Extract(ctx context.Context, raw []byte, opts ExtractOptions) (ExtractResult, error) {
	if len(raw) == 0 {
		return ExtractResult{
			Text:     "",
			Metadata: map[string]string{"ocr_engine": "none"},
			Skipped:  true,
			Reason:   "empty input",
		}, nil
	}

	language := e.resolveLanguage(opts)

	// Try PaddleOCR first.
	if CheckBinaryAvailable(ctx, e.cfg.PaddleOCRBinary) {
		result, err := e.extractWithPaddleOCR(ctx, raw, language, opts)
		if err == nil {
			return result, nil
		}
		e.cfg.Logger.Warn("paddleocr extraction failed, falling back to tesseract",
			"error", err,
			"file", opts.FileName,
		)
	}

	// Fallback to Tesseract.
	if CheckBinaryAvailable(ctx, e.cfg.TesseractBinary) {
		result, err := e.extractWithTesseract(ctx, raw, language, opts)
		if err == nil {
			return result, nil
		}
		e.cfg.Logger.Warn("tesseract extraction failed",
			"error", err,
			"file", opts.FileName,
		)
		return ExtractResult{}, fmt.Errorf("ingest: image extraction failed with both OCR engines: %w", err)
	}

	return ExtractResult{}, fmt.Errorf("ingest: no OCR engine available (tried %s, %s)", e.cfg.PaddleOCRBinary, e.cfg.TesseractBinary)
}

// ExtractStream buffers the reader and delegates to Extract.
func (e *ImageExtractor) ExtractStream(ctx context.Context, reader io.Reader, opts ExtractOptions) (ExtractResult, error) {
	var limitReader io.Reader = reader
	if opts.MaxBytes > 0 {
		limitReader = io.LimitReader(reader, opts.MaxBytes)
	}
	raw, err := io.ReadAll(limitReader)
	if err != nil {
		return ExtractResult{}, fmt.Errorf("ingest: reading image stream: %w", err)
	}
	return e.Extract(ctx, raw, opts)
}

// resolveLanguage determines the OCR language parameter from options
// or falls back to the configured default.
func (e *ImageExtractor) resolveLanguage(opts ExtractOptions) string {
	if opts.Encoding != "" {
		return opts.Encoding
	}
	return e.cfg.DefaultLanguage
}

// writeTempFile writes raw bytes to a secure temporary file and
// returns the file path. The caller is responsible for removing the
// file.
func writeTempFile(raw []byte, pattern string) (string, error) {
	f, err := os.CreateTemp("", pattern)
	if err != nil {
		return "", fmt.Errorf("ingest: creating temp file: %w", err)
	}

	path := f.Name()

	if _, err := f.Write(raw); err != nil {
		f.Close()
		os.Remove(path)
		return "", fmt.Errorf("ingest: writing temp file: %w", err)
	}

	if err := f.Close(); err != nil {
		os.Remove(path)
		return "", fmt.Errorf("ingest: closing temp file: %w", err)
	}

	return path, nil
}

// extractWithPaddleOCR invokes PaddleOCR via subprocess. PaddleOCR
// does not accept stdin, so the image is written to a temporary file.
func (e *ImageExtractor) extractWithPaddleOCR(ctx context.Context, raw []byte, language string, opts ExtractOptions) (ExtractResult, error) {
	ext := extensionFromOpts(opts)
	tmpPath, err := writeTempFile(raw, "memory-ocr-*"+ext)
	if err != nil {
		return ExtractResult{}, err
	}
	defer os.Remove(tmpPath)

	// Resolve to absolute path to prevent path traversal.
	absPath, err := filepath.Abs(tmpPath)
	if err != nil {
		return ExtractResult{}, fmt.Errorf("ingest: resolving temp path: %w", err)
	}

	paddleLang := mapLanguageToPaddleOCR(language)

	args := []string{
		"--image_dir", absPath,
		"--use_angle_cls", "true",
		"--lang", paddleLang,
		"--type", "ocr",
	}

	result, err := RunSubprocess(ctx, e.cfg.PaddleOCRBinary, args, nil, SubprocessOptions{
		Timeout: e.cfg.Timeout,
	})
	if err != nil {
		return ExtractResult{}, fmt.Errorf("ingest: paddleocr: %w", err)
	}

	if result.ExitCode != 0 {
		return ExtractResult{}, fmt.Errorf("ingest: paddleocr exited with code %d: %s", result.ExitCode, result.Stderr)
	}

	text, confidence := parsePaddleOCROutput(string(result.Stdout))

	return ExtractResult{
		Text:        text,
		ContentType: opts.ContentType,
		Metadata: map[string]string{
			"ocr_engine":     "paddleocr",
			"ocr_confidence": formatConfidence(confidence),
			"ocr_language":   paddleLang,
		},
		Language:   language,
		Confidence: confidence,
	}, nil
}

// extractWithTesseract invokes Tesseract via subprocess. The image is
// written to a temporary file since Tesseract's stdin support varies
// across versions.
func (e *ImageExtractor) extractWithTesseract(ctx context.Context, raw []byte, language string, opts ExtractOptions) (ExtractResult, error) {
	ext := extensionFromOpts(opts)
	tmpPath, err := writeTempFile(raw, "memory-ocr-*"+ext)
	if err != nil {
		return ExtractResult{}, err
	}
	defer os.Remove(tmpPath)

	absPath, err := filepath.Abs(tmpPath)
	if err != nil {
		return ExtractResult{}, fmt.Errorf("ingest: resolving temp path: %w", err)
	}

	tesseractLang := mapLanguageToTesseract(language)

	args := []string{
		absPath,
		"stdout",
		"-l", tesseractLang,
		"--oem", "3",
		"--psm", "3",
	}

	result, err := RunSubprocess(ctx, e.cfg.TesseractBinary, args, nil, SubprocessOptions{
		Timeout: e.cfg.Timeout,
	})
	if err != nil {
		return ExtractResult{}, fmt.Errorf("ingest: tesseract: %w", err)
	}

	if result.ExitCode != 0 {
		return ExtractResult{}, fmt.Errorf("ingest: tesseract exited with code %d: %s", result.ExitCode, result.Stderr)
	}

	text := strings.TrimSpace(string(result.Stdout))

	return ExtractResult{
		Text:        text,
		ContentType: opts.ContentType,
		Metadata: map[string]string{
			"ocr_engine":   "tesseract",
			"ocr_language": tesseractLang,
		},
		Language: language,
	}, nil
}

// parsePaddleOCROutput parses PaddleOCR's JSON output format.
// PaddleOCR outputs an array of [text, confidence] pairs. The text
// blocks are concatenated in reading order and the average confidence
// is computed.
func parsePaddleOCROutput(output string) (string, float64) {
	// PaddleOCR outputs results line by line. The actual text results
	// follow a pattern of JSON-like arrays. Try JSON parsing first,
	// then fall back to line-by-line text extraction.
	lines := strings.Split(output, "\n")
	var textBlocks []string
	var totalConfidence float64
	var blockCount int

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		// Try parsing as JSON array: [["text", confidence], ...]
		var entries [][]json.RawMessage
		if json.Unmarshal([]byte(line), &entries) == nil {
			for _, entry := range entries {
				if len(entry) < 2 {
					continue
				}
				var text string
				var conf float64
				if json.Unmarshal(entry[0], &text) == nil && json.Unmarshal(entry[1], &conf) == nil {
					textBlocks = append(textBlocks, text)
					totalConfidence += conf
					blockCount++
				}
			}
			continue
		}

		// Try parsing individual result line: ("text", confidence)
		// PaddleOCR CLI sometimes outputs: ('text', 0.95)
		if strings.HasPrefix(line, "(") && strings.HasSuffix(line, ")") {
			inner := line[1 : len(line)-1]
			// Find the last comma to split text from confidence.
			lastComma := strings.LastIndex(inner, ",")
			if lastComma > 0 {
				textPart := strings.TrimSpace(inner[:lastComma])
				confPart := strings.TrimSpace(inner[lastComma+1:])

				// Strip quotes from text part.
				textPart = stripQuotes(textPart)

				if conf, err := strconv.ParseFloat(confPart, 64); err == nil {
					textBlocks = append(textBlocks, textPart)
					totalConfidence += conf
					blockCount++
				}
			}
		}
	}

	text := strings.Join(textBlocks, "\n")
	avgConfidence := 0.0
	if blockCount > 0 {
		avgConfidence = totalConfidence / float64(blockCount)
	}

	return text, avgConfidence
}

// stripQuotes removes surrounding single or double quotes from a
// string.
func stripQuotes(s string) string {
	if len(s) < 2 {
		return s
	}
	first, last := s[0], s[len(s)-1]
	if (first == '\'' && last == '\'') || (first == '"' && last == '"') {
		return s[1 : len(s)-1]
	}
	return s
}

// extensionFromOpts determines the file extension from the extract
// options, defaulting to ".png".
func extensionFromOpts(opts ExtractOptions) string {
	extensionMap := map[string]string{
		"image/png":  ".png",
		"image/jpeg": ".jpg",
		"image/tiff": ".tiff",
		"image/bmp":  ".bmp",
		"image/webp": ".webp",
	}

	if ext, ok := extensionMap[opts.ContentType]; ok {
		return ext
	}

	if opts.FileName != "" {
		ext := filepath.Ext(opts.FileName)
		if ext != "" {
			return ext
		}
	}

	return ".png"
}

// mapLanguageToPaddleOCR maps ISO 639-1 language codes to PaddleOCR's
// language parameter format.
var paddleOCRLanguageMap = map[string]string{
	"en": "en",
	"zh": "ch",
	"ja": "japan",
	"ko": "korean",
	"fr": "fr",
	"de": "german",
	"es": "es",
	"pt": "pt",
	"it": "it",
	"ru": "ru",
	"ar": "ar",
	"hi": "hi",
	"ta": "ta",
	"te": "te",
}

func mapLanguageToPaddleOCR(lang string) string {
	normalised := strings.ToLower(strings.TrimSpace(lang))
	if mapped, ok := paddleOCRLanguageMap[normalised]; ok {
		return mapped
	}
	return normalised
}

// mapLanguageToTesseract maps ISO 639-1 codes to Tesseract's 3-letter
// language codes.
var tesseractLanguageMap = map[string]string{
	"en": "eng",
	"zh": "chi_sim",
	"ja": "jpn",
	"ko": "kor",
	"fr": "fra",
	"de": "deu",
	"es": "spa",
	"pt": "por",
	"it": "ita",
	"ru": "rus",
	"ar": "ara",
	"hi": "hin",
	"ta": "tam",
	"te": "tel",
}

func mapLanguageToTesseract(lang string) string {
	normalised := strings.ToLower(strings.TrimSpace(lang))
	if mapped, ok := tesseractLanguageMap[normalised]; ok {
		return mapped
	}
	return "eng"
}

// formatConfidence formats a float64 confidence value as a string with
// 4 decimal places.
func formatConfidence(conf float64) string {
	return strconv.FormatFloat(conf, 'f', 4, 64)
}
